lstmcpipe/workflow_management.py

```diff
@@ -51,146 +52,45 @@ def batch_r0_to_dl1(input_dir, conf_file, prod_id, particles_loop, source_env, g
     all_jobids_from_r0_dl1_stage = []
 
     print("\n ==== START {} ==== \n".format('batch r0_to_dl1_workflow'))
+    
+    def start_r0():
```

Comment:
Not sure if I understand you correctly.
The function is called in a loop over the particles where the input dir is set just like it was before (unless I made a mistake of course). It should be the same code being run collected in a function instead of being duplicated in the file

Reply:
I think it is highly surprising to rely on the surrounding scope to update the internal variables here, that's not really what closures are made for. Just give what changes as an argument to the function and then it does not need to be a closure at all.

Closures are only easy to reason about if the values they capture are not mutated outside the closure.

Reply:
> The function is called in a loop over the particles where the input dir is set just like it was before

Sorry Lukas, now I see what you did. The code indeed should work as before.

Reply:
> I think it is highly surprising to rely on the surrounding scope to update the internal variables here, that's not really what closures are made for. 

> Closures are only easy to reason about if the values they capture are not mutated outside the closure.

I agree @maxnoe, however I'm not sure it will simplify the code if you pass all the arguments to the final function. Maybe it will make it clearer, I agree though. Let me explain a bit.

Currently what's going on is that the the common paths are created when the pipeline is launched, i.e.
`$ROOT_PATH/$STAGE/$PROD_DATE/$PARTICLE/$ZENITH/$POINTING/$PROD_ID/$GAMMA_OFF

If we pass the needed args to `start_r0`, we would need to pass $PARTICLE, $ZENITH, $POINTING, $PROD_ID and $GAMMA_OFF. 

Currently the only vars that we need to manage are $PARTICLE and $GAMMA_OFF. 
The code is less readable, I agree, but avoids passing 6 args through 2 or 3 functions. 

(Of course there would be much nicer ways of code this, I'm aware)

Reply:
Just make a `namedtuple` (maybe `ProductionInfo` ?) and pass that around?

Reply:
Let me have a look to it. I'm not super familiarised with `namedtuple` (not sure how we would manage the gamma_off cases). Thanks in any case for the suggestion, Max.